### PR TITLE
Improve affordance of layer item

### PR DESCRIPTION
**Pull request by matallo**
fixes: #11359 

this PR makes the changes requested by @CartoDB/design to improve the affordance of both basemaps and layer items in the layers list

now all the card target is clickable, but:

- the context menu
- the source layer
- and the title (in case of doubleclick), as shown in the graphic below

captura de pantalla 2017-01-25 a las 9 28 59

also, changes in interaction styles have been made, whenever the mouse hovers the card the title is underlined, unless you hover the source layer which will be underlined instead (the destination changes), as in the next screenshots

01 initial state
02 hover layer
03 hover dataset

**Acceptance**

- [x] Interact with any target of the card (hover)
- [x] Click and select options from the context menu
- [x] Show/hide layer with the corresponding icon
- [x] Go to style layer with the corresponding icon (the style tab should be selected)
- [x] Enter to edit the layer (the data tab should be selected)
- [x] Double click the title to edit
- [x] Delete layer should work from layers list view (with the options).
- [x] Delete layer should work from layer view (with the options).
- [x] Drag the layer.
- [x] Drag an analysis.

if there is no geometry the style icon should not appear

some changes have been made on the stack layer, so further check when passing analysis params the analysis tab is selected

### CODE REVIEW

**Comment by saleiva** [comment]
very good! I'd like to help with acceptance, let me know where to test!

cc/ @noguerol

**Comment by matallo** [comment]
there's just the doubt about the destination of the table link, like @xavijam suggested I'd go to the table url

**Comment by saleiva** [comment]
Uhm... just looked at it on FS, some users are clicking on it for getting to the layer details so not sure such a big element should link outside the page. My proposal is to link from there to the DATA tab within the layer.

**Comment by xavijam** [comment]
Data tab for the win then!

**Comment by matallo** [comment]
made the suggested changes by @saleiva and ready to move to CR

File: app/assets/stylesheets/editor-3/layers-list.css.scss [hunk 1]
```diff
@@ -107,8 +105,18 @@
   left: -20px;
   opacity: 0;
 }
-.Editor-ListLayer-item:hover .Editor-ListLayer-dragIcon {
-  opacity: 1;
+.Editor-ListLayer-item:hover {
```
Comment (javierarce) [inline app/assets/stylesheets/editor-3/layers-list.css.scss]: Also this☝️

Comment (matallo) [inline app/assets/stylesheets/editor-3/layers-list.css.scss]: this warning remained after doing the change :S anyway, I moved the hover state nested to the `.Editor-ListLayer-item` selector

File: lib/assets/core/javascripts/cartodb3/editor/layers/analysis-views/default-layer-analysis-view.js [hunk 2]
```diff
@@ -27,8 +33,18 @@ module.exports = CoreView.extend({
...
+
+    this._stateModel = new Backbone.Model({
+      highlighted: false
+    });
+
+    this._initBinds();
+  },
+
+  _initBinds: function () {
+    this._stateModel.on('change:highlighted', this._toggleHover, this);
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/analysis-views/default-layer-analysis-view.js]: `listenTo` all the things! 👍

File: lib/assets/core/javascripts/cartodb3/editor/editor-map-view.js [hunk 3]
```diff
@@ -91,7 +91,8 @@ module.exports = CoreView.extend({
             });
           case 'layer-content':
             var layerDefinitionModel = opts[0];
-            var analysisPayload = opts[2];
+            var selectedTabItem = opts[2];
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/editor-map-view.js]: agree, the current solution is not ideal as we deal differently with `selectedTabItem` depending in the case

(just see a couple of lines before in https://github.com/CartoDB/cartodb/blob/7634e3cf01c67cb15bbd60ddb01c9e5d5a8f6468/lib/assets/core/javascripts/cartodb3/editor/editor-map-view.js#L56)

`var selectedTabItem = opts[0] || 'layers';`

we are always sending `selectedTabItem` with a value vs https://github.com/CartoDB/cartodb/blob/11359-affordance-layer/lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js#L79 where if  `selectedTabItem` is not defined we default to `data`

I do agree we need to address this issue at some point but first consolidate how selectedtab should work on the `StackView`

@xavijam ^^^

File: lib/assets/core/javascripts/cartodb3/editor/editor-map-view.js [hunk 4]
```diff
@@ -226,8 +228,8 @@ module.exports = CoreView.extend({
     }.bind(this));
     this._modals.onDestroyOnce(function (analysisFormAttrs) {
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/editor-map-view.js]: not straight forward as we are making use of `layerDefinitionModel` variable previously defined and we can't just `this._layerDefinitionModel`

will leave it like this in the meantime but after my pairing with @xavijam I feel we're improving this soon

File: lib/assets/core/javascripts/cartodb3/editor/layers/analysis-views/source-layer-analysis-view.js [hunk 5]
```diff
@@ -79,8 +88,25 @@ module.exports = CoreView.extend({
   },
 
   _initBinds: function () {
+    this._stateModel.on('change:highlighted', this._toggleHover, this);
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/analysis-views/source-layer-analysis-view.js]: 👍

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js [hunk 6]
```diff
@@ -69,12 +70,13 @@ module.exports = CoreView.extend({
...
 
     var self = this;
+    var selectedTabItem = this.options.selectedTabItem;
     var analysisPayload = this.options.analysisPayload;
 
     var tabPaneTabs = [{
       label: _t('editor.layers.menu-tab-pane-labels.data'),
       name: 'data',
-      selected: !analysisPayload,
+      selected: (!selectedTabItem || selectedTabItem && selectedTabItem === 'data'),
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js]: ha, I actually did it like this first but I thought it was more legible, applying your suggestion 👍

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js [hunk 7]
```diff
@@ -138,7 +143,11 @@ module.exports = CoreView.extend({
     }, {
       label: _t('editor.layers.menu-tab-pane-labels.style'),
       name: 'style',
+      selected: !!(selectedTabItem && selectedTabItem === 'style'),
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js]: totally not 👍

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-views/data-layer-view.js [hunk 8]
```diff
@@ -172,12 +179,39 @@ module.exports = CoreView.extend({
...
+    event.stopPropagation();
+
+    var nodeId = event.currentTarget && event.currentTarget.dataset.analysisNodeId;
+    if (!nodeId) throw new Error('missing data-analysis-node-id on element to edit analysis node, the element was: ' + event.currentTarget.outerHTML);
+
+    var nodeDefModel = this._analysisDefinitionNodesCollection.get(nodeId);
+    var layerDefModel = this._layerDefinitionsCollection.findOwnerOfAnalysisNode(nodeDefModel);
+    if (!layerDefModel) throw new Error('no owning layer found for node ' + nodeId);
+
+    this._stackLayoutModel.nextStep(layerDefModel, 'layer-content', 'analyses', nodeId);
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-views/data-layer-view.js]: uhm, I'm not feeling confident about this change in this specific PR, as we should do something similar in other parts of the app the the `stackLayoutModel` steps, open to suggestions tho

Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-views/data-layer-view.js]: You are totally right, let's leave it like it's now :)

File: lib/assets/core/javascripts/cartodb3/editor/layers/layers-view.js [hunk 9]
```diff
@@ -28,6 +28,7 @@ module.exports = CoreView.extend({
     if (!opts.editorModel) throw new Error('editorModel is required');
     if (!opts.userActions) throw new Error('userActions is required');
     if (!opts.stateDefinitionModel) throw new Error('stateDefinitionModel is required');
+    if (!opts.widgetDefinitionsCollection) throw new Error('widgetDefinitionsCollection is required');
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layers-view.js]: done, and also in the files below

https://github.com/CartoDB/cartodb/pull/11534/files#diff-23932f0ef5482ef6ae90116839c5b72bR44
https://github.com/CartoDB/cartodb/pull/11534/files#diff-78e7917748a5088168889c8b2a7339b3R47

**Comment by javitonino** [comment]
✅ Looking good. Functionally, everything works as expected. Just a couple of minor comments about the design:

- I think hiding the layer should also fade the X analyses / Y widgets text. It looks weird as this text isn't greyed out on hiding the layer:
screenshot_20170216_101349
- First time I tried to rename a layer, I failed. It turns out my mouse was outside the title area (just a little below), but as the title was underlined, it felt like I was inside the active area. Minor, and I got it right at the second try.

@saleiva This is deployed to ded07, if you want to test it.

**Comment by matallo** [comment]
good points, regarding the number of analyses this is already happening with the add analyses link and I decided to leave it as is

**Comment by arianaescobar** [comment]
+1 on your first point @javitonino, those should definitely fade away. I actually opened a ticket about a few months ago but I can't find it :(

**Comment by xavijam** [comment]
I'll fix it, no problem VAMOS!

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-views/data-layer-view.js [hunk 10]
```diff
@@ -293,7 +327,7 @@ module.exports = CoreView.extend({
         layerModel: self.model,
         modalModel: modalModel,
         visDefinitionModel: this.visDefinitionModel,
-        widgetDefinitionsCollection: this.widgetDefinitionsCollection
+        widgetDefinitionsCollection: this._widgetDefinitionsCollection
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-views/data-layer-view.js]: I've found a problem with the delete-confirmation-modal after this change. Fixing it! no prob!

**Comment by xavijam** [comment]
Fixed that:

screen shot 2017-02-16 at 12 22 09

And also changed one thing that squeaked me. Basically we were using "window" variables inside delete-layer-confirmation-modal view creation, and I don't like that, because it could cause us problems in the near future (cc @javierarce @nobuti @matallo @ivanmalagon). For example:

- https://github.com/CartoDB/cartodb/pull/11534/files#diff-78e7917748a5088168889c8b2a7339b3L295
- https://github.com/CartoDB/cartodb/pull/11534/files#diff-fd1f8821a1e75fb5d9f831ecaab9228eL313

**Comment by xavijam** [comment]
Moving to CR again, but I encourage to check only new things 💋 .

**Comment by xavijam** [comment]
Added delete layer options in acceptance list. Thanks @javierarce

**Comment by javierarce** [comment]
Ah, good idea, thanks!

**Comment by javitonino** [comment]
Found a funny one. After dragging a layer, you enter into the layer style view. Can we disable the new click handlers when dragging? It makes dragging layers a confusing experience. Added it to acceptance steps.

I tested all actions I could find in the layer list view, as well as the main ones in the layer detail view and everything worked.

**Comment by matallo** [comment]
taking that one @javitonino thanks for the thorough test

**Comment by matallo** [comment]
this is already happening for Firefox in production, shall we open a new issue and move on with this PR?

**Comment by javitonino** [comment]
The ticket: https://github.com/CartoDB/cartodb/issues/11558

**Comment by matallo** [comment]
🚀

**Pull request merged**